### DISCUSSION TITLE
`ToolsTreePackages=` vs `BuildPackages=`

### DISCUSSION THREAD

**Discussion by septatrix** [systemd/mkosi#2390] [Q&A] [answered]
Is there any guidelines on when to prefer which?

For example `pip` (for installing Python packages not shipped by the distro) could be reasonably in either the tools tree or build tree - even more so as it is not clear whether this should be done in `mkosi.build` or `mkosi.prepare`. Another example which might further blur the line could be `nodejs` and `npm` (or similar tool), in addition to installing packages these can often be used to "compile"/transcribe source files into standalone bundles.

I am not sure if there might be better distinction when compiling e.g. C where header files play a role but to me it is not clear what to use when - or is it really that in my case it does not really matter?

---

PS: One advantage of installing dependencies in `mkosi.prepare` would be to have it included in the cached image and being able to run the build without network access but on the other hand one does not have access to the build sources. The documentation actually recommends doing this inside the prepare stage. However, in the case of Python and NodeJS, the dependencies are commonly contained in files like `requirements.txt` or `package-lock.json` but one cannot access build sources in the prepare step... So again one could argue for both things here.

Maybe this is also a sign that mkosi modeling of building software reaches its limits here as it is mostly focus on C where the dependencies are usually bundled by the distro and slowly changing. Whereas for Python, NodeJS, and also kinda Rust, this is less the case and a two-stage build might be preferred, `mkosi.build fetch` and `mkosi.build build`. Or maybe some way to a subset of build sources into the prepare step? Vendoring them inside the same repository where the mkosi instructions are is not always feasible.

**Comment by daandemeyer** [comment, marked as answer]
@septatrix You do have access to the sources in the prepare script, every script has access to the sources.

You use build packages if you want/need to chroot into the image to install stuff. Everything that is installed as a tools tree package needs to have a `--root` option or equivalent to be able to operate on the image.

**Reply by septatrix** [reply]
Oh yes indeed, hm, I must have gotten fed up with pip and pnpm and their support for installing to a custom `--root`. I gave it another go and everything seems to work now, thanks

**Reply by septatrix** [reply]
Ah I think I remember what my problem back then was. I require some nodejs packages for building but they should not be included in the final image. Previously `pnpm` would install them in the build sources which was rather annoying as they were owned by a different user. Now I just use `BuildSourcesEphemeral` which works great for this. I likely could also use `setpriv` but I think I prefer the former approach as it "just work TM".
